applications/staex-iod.md

```diff
@@ -0,0 +1,326 @@
+4. Provisioner gets tokens from smart contract.
+5. Provisioner starts to distribute data for this user.
+
+##### Data access smart contract
+
+To implement flow below we will implement a kind of multisig ink! smart contract with three methods:
+
+- Request - user requests for particular data access and sends required amount of tokens.
+- Accept - device accepts request and gets required amount of tokens.
+- Reject - device rejects request and tokens will be returned to user.
+
+Device can reject request because of:
+
+- Requested data is not available at the moment.
+- Device has high amount of request (highload).
+
+This contract will use DID smart contract to check that user sent required amount of tokens otherwise reject automatically.
+
+##### Data distribution
+
+To distribute data for exact user we want to use IPFS and it is fully open source, distributed and IoT devices don't need to have public IP.
+
+To make device root folder static in IPFS network we will use IPNS. With IPNS user which requested for data can look for exact place in IPFS to find it. So IPFS structure will be following:
+
+```shell
+/ipfs/<device public key>/<user public key>/<file name>
+```
+
+With such structure every user can know where requested data is located.
+
+When provisioner gets tokens from smart contract it will start distributing process. To prevent data reading from IPFS by users which didn't pay for it, provisioner will encrypt data before upload. To generate encrypt key we will use Diffie–Hellman key exchange. With this algorithm we can generate encrypt/decrypt key without transferring it by the network. So device private key + user public key and device public key + user private key can generate exact the same shared secret key which provisioner will use for encryption.
+
+After getting data from database and encrypting it provisioner will save and pin data in IPFS and notify user through IPFS pub/sub protocol.
+
+##### Data accessing
+
+To access decrypted data automatically user can start provisioner on their own laptop or server. Provisioner will start local IPFS node and subscribe for IPFS events and start sync files from IPFS to local database. Also it will generate shared secret key and decrypt all data while sync process. 
+
+Also provisioner will start simple server to access data from UI through HTTP. UI will be simple and user can see a table with synced data. Each row will have some metadata like IoT device public key, data created time and data itself. Such UI also will be available on device to see produced data.
+
+##### Data access prolongation
+
+For example data cost 1 token per 1KB. If device saved to IPFS 1KB of data it will stop data distribution. If user doesn't want such data access prevents it should put more tokens on smart contract on data request. When provisioner almost done with 1KB it will try to get more tokens from smart contract and in case of success continue to distribute data.
```

Comment:
Which token are you referring to here? Are you saying it would cost 1 KSM for 1KB?

Reply:
I think it can be any token like DOT/KSM/SDN/..., but for this phase of the project I assume it will be SBY.